sdk/nodejs/resource.ts

```diff
@@ -421,12 +417,59 @@ To manipulate the value of this Output, use '.apply' instead.`);
     }
 }
 
-export function output<T>(cv: Input<T>): Output<T>;
-export function output<T>(cv: Input<T> | undefined): Output<T | undefined>;
-export function output<T>(cv: Input<T | undefined>): Output<T | undefined> {
-    // outputs created from simply inputs are always stable.
-    return Output.isInstance<T | undefined>(cv) ? cv :
-        new Output<T | undefined>(new Set<Resource>(), Promise.resolve(cv), Promise.resolve(true));
+/**
+ * [output] takes in a object and creates a deep clone of it with all inner Outputs and Promises
+ * unwrapped.  i.e. the object returned will point at the values inside those Outputs and Promises.
+ * This process works transitively over the entire value graph and will see inside of Arrays and
+ * Objects.
+ *
+ * The resultant awaited value of this function will be an Output containing the final completely
+ * unwrapped object, as well as all [Resource]s that were encountered along the way while unwrapping
+ * (not including Promise boundaries). With this, the result can then be transformed using
+ * [Output.apply] as usual, and the result of that can be passed anywhere that needs such a value
+ * and also wants to keep track of dependent [Resource]s.  The expected way to use this function is
+ * like:
+ *
+ * ```ts
+ *      var transformed = pulumi.unwrap(someVal).apply(unwrapped => {
+ *          // Do whatever you want now.  'unwrapped' will contain no outputs/promises inside
+ *          // here, so you can easily do whatever sort of transformation is most convenient.
+ *      });
+ *
+ *      // Result can be passed to another Resource.  The dependency information will be
+ *      // properly maintained.
+ *      var someResource = new SomeResource(name, { data: transformed ... });
+ * ```
+ */
+export function output<T>(val: Input<T>): Output<Unwrap<T>>;
+export function output<T>(val: Input<T> | undefined): Output<Unwrap<T | undefined>>;
+export function output<T>(val: Input<T | undefined>): Output<Unwrap<T | undefined>> {
```

Comment:
I know this is unchanged - but I'm unclear on why these set of signatures was every required.  The second seems to unnecessarily overlap with the first, and the implementation signture seems to unnecessarily include things that aren't valid in either of the exposed signatures.  Not a big deal - but if we can clean up - may as well while we are touching this.

Reply:
Yeah, something seems wonky here.  I'll see if we can clean this up.

Reply:
Ok.  So i looked at this, and i think this is waht we want.  First, let's completely ignore the impl signatures, it's irrelevant.  So we have these two sigs:

```ts
output<T>(val: Input<T>): Output<T>;
output<T>(val: Input<T> | undefined): Output<T | undefined>;
```
(i'm leaving out 'Unwrap' as it doesn't help with the explanation, and it keeps things simpler).

Now, say you only had the first form.  But say you had the following type somewhere:

```ts
interface ClusterArgs {
    name: Input<string>;
    isPublic?: Input<boolean>;
    // ...
}

const c: ClusterArgs = //...
const val = output(c.isPublic).apply(...);
```

If you only had the first 'output' form this code would not be legal.  TS would complain that you were passing an ```Input<boolean> | undefined``` to something that only took an ```Input<boolean>```.

Now, say we only had the second form of ```output```.  That would not be a great situation either in the following code:

```ts
const c: ClusterArgs = //...
const val = output(c.name).apply(name => /* here name is possibly undefined */);
```

Now, every time you call .apply, you might have to deal with something 'undefined' even if there was no way that would happen.

The two forms allow us to express this duality cleanly.  If you have a non-optional Input property, then you call the first form, and you get an Output which gives you a non-optional inner value.  If you have an optional property, you call the second form, and you get a possibly undefined inner value (as expected).  

In *both* cases you get an 'Output' though, allowing you to always feel safe doing ```output(val).apply(...)```.

--

Does that make sense @lukehoban ?  If you feel otherwise, let me know what you're thinking.